forest/willow/log_stats.py

```diff
@@ -437,121 +437,104 @@ def log_stats_main(
-                                    "of individual cardinalities for at "
-                                    "least one time interval. This error "
-                                    "comes from an issue with the code,"
-                                    " not an issue with the input data"
-                                    )
-
-                        write_all_summaries(bid, stats_pdframe, output_folder)
-
-                    logger.info(
-                        "Summary statistics obtained. Finished."
-                    )
+    logger.info("Summary statistics obtained. Finished.")
 
-            except Exception as err:
-                logger.error(
-                    "An error occurred when processing the data: %s", err
-                )
+
+def log_stats_inner(
+    beiwe_id: str,
+    output_folder: str,
+    study_folder: str,
+    frequency: Frequency,
+    tz_str: str,
+    time_start: Optional[List] = None,
+    time_end: Optional[List] = None,
+):
+    """ Inner functionality of log_stats_main """
+    # read data
+    text_data, text_stamp_start, text_stamp_end = read_data(
+        beiwe_id, study_folder, "texts", tz_str, time_start, time_end
+    )
+    call_data, call_stamp_start, call_stamp_end = read_data(
+        beiwe_id, study_folder, "calls", tz_str, time_start, time_end
+    )
+
+    # give up early if there is no data
+    if text_data.shape[0] <= 0 and call_data.shape[0] <= 0:
+        logger.info(f"There was no data for participant {beiwe_id}")
+        return
+
+    # stamps from call and text should be the stamp_end
+    logger.info("Data imported ...")
+    stamp_start = min(text_stamp_start, call_stamp_start)
+    stamp_end = max(text_stamp_end, call_stamp_end)
+
+    # process the data
+    stats_pdframe = comm_logs_summaries(
+        text_data, call_data, stamp_start, stamp_end, tz_str, frequency
+    )
+
+    # num_uniq_individuals_call_or_text is the cardinality of the union of
+    # several sets. It should should always be at least as large as the
+    # cardinality of any one of the sets, and it should never be larger than
+    # the sum of the cardinalities of all of the sets. (it may be equal if all
+    # the sets are disjoint)
+    num_uniq_column = "num_uniq_individuals_call_or_text"  # legibility hax.
+    sum_all_set_cols = pd.Series([0]*stats_pdframe.shape[0])
+    for column in [
+        "num_s_tel", "num_r_tel", "num_in_caller",
+        "num_out_caller", "num_mis_caller"
+    ]:
+        sum_all_set_cols += stats_pdframe[column]
+        if (stats_pdframe[num_uniq_column] < stats_pdframe[column]).any():
+            logger.error(
+                "Error: "
+                "num_uniq_individuals_call_or_text was found to be less than "
+                "%s for at least one time interval. This error comes from an "
+                "issue with the code, not an issue with the input data." %
+                column
+            )
```

Comment:
We use `logger.info("%s", var)` style in log strings (https://forest.beiwe.org/en/latest/development.html#python)